File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> ans;
        if (s.length() == 0 || words.empty()) {
            return ans;
        }
        unordered_map<string, int> hash;
        for (auto& w : words) {
            hash[w]++;
        }

        int len = words[0].size();
        int n = s.length();
        for (int start = 0; start < len; start++) {
            unordered_map<string, int> window;
            int left = start;
            int count = 0;
            for (int r = start; r <= n - len; r += len) {
                string sub = s.substr(r, len);
                auto it = hash.find(sub);
                if (it == hash.end()) {
                    window.clear();
                    count = 0;
                    left = r + len;
                    continue;
                }

                window[sub]++;
                count++;
                while(window[sub] > hash[sub]) {
                    string pop = s.substr(left, len);
                    window[pop]--;
                    count--;
                    left += len;
                }

                if (count == words.size()) {
                    ans.push_back(left);
                }
            }
        }
        return ans;
    }
};

File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.go
func findSubstring(s string, words []string) []int {
    ans := make([]int, 0)
    if len(s) == 0 || len(words) == 0 {
        return ans
    }

    word2int := make(map[string]int, 0)
    for _, w := range words {
        word2int[w]++
    }

    length := len(words[0])
    n := len(s)
    for offset := 0; offset < length; offset++ {
        windows := make(map[string]int)
        left := offset
        size := 0
        for i := offset; i <= n - length; i+=length {
            sub := s[i:i+length]
            count, ok := word2int[sub]
            if !ok {
                windows = make(map[string]int)
                size = 0
                left = i + length
                continue
            }
            windows[sub]++
            size++
            for windows[sub] > count {
                move := s[left:left+length]
                windows[move]--
                size--
                left += length
            }

            if size == len(words) {
                ans = append(ans, left)
            }
        }
    }
    return ans
}